Allow comma and space delimited values

scowcron opened the pull request:
According to the libconfig grammar, comma and space are valid
separators for settings as well as semicolons. Allow them.

setting = name (":" | "=") value (";" | "," | empty)

scowcron committed:
Allow comma and space delimited values

According to the libconfig grammar, comma and space are valid
separators for settings as well as semicolons. Allow them.

setting = name (":" | "=") value (";" | "," | empty)

filipegoncalves commented:
Thanks for your contribution.

The crate's documentation (http://filipegoncalves.github.io/rust-config/config/) should be the official source for the grammar; in particular, the entry that describes settings is:

```
setting -> __ name __ (":"|"=") __ value __ ";" __
```

Where `__` is any number of blanks or block comments. It explicitly shows that settings should always be terminated by `;`.

Could you please elaborate on the rationale behind these changes? Is there a specific use case or situation where you find this beneficial?

scowcron commented:
I have a few configuration files I share between programs. They're all formatted without any semicolons to speak of, just separated by newlines.

This is my first foray into rust, but I thought I'd bring them with, and saw the libconfig style library and thought it was a done deal. Unfortunately the built-in error reporting still needs some work and I couldn't find any documentation for the crate (your page returns a 404), so I was left to the source. This patch is where I landed.

It would be easy enough to simply add semicolons to my configurations since i haven't distributed any of the programs that use them and that would make my files compatible with both libraries. However, I chose to omit the semicolons for aesthetic reasons, and my preference would be not to have to add them.

filipegoncalves commented on src/reader.rs:
```diff
@@ -38,7 +38,7 @@ use error::{from_io_err, from_parse_err};
 /// use config::error::ConfigErrorKind;
 ///
-/// let sample_conf = "windows=\n";
+/// let sample_conf = "windows=NO SPACE\n";
```
Why `windows=NO SPACE`? A configuration containing `windows=` should still fail to parse.

filipegoncalves commented on src/reader.rs:
```diff
@@ -127,7 +127,7 @@ pub fn from_file(path: &Path) -> Result<Config, ConfigError> {
 /// use config::error::ConfigErrorKind;
 ///
-/// let parsed = from_str("windows=NO\n");
+/// let parsed = from_str("windows=NO SPACE\n");
```
I understand that with this patch `windows=NO\n` is a valid line and as such it breaks the example. But replacing it with `windows=NO SPACE` seems a bit misleading to me, maybe we should find a better example? I think this would make the reader confused, don't you agree?

filipegoncalves commented on src/reader.rs:
```diff
@@ -195,7 +195,7 @@ mod test {
     #[test]
     fn conf_from_str_parse_err() {
-        let parsed = from_str("windows=NO\nlinux=true;\n");
+        let parsed = from_str("windows=NO SPACE\nlinux=true;\n");
```
Ditto.

filipegoncalves commented on src/reader.rs:
```diff
@@ -223,7 +223,7 @@ mod test {
     #[test]
     fn conf_from_stream_parse_err() {
-        let sample_conf = "windows=NO\nlinux = true;\n";
+        let sample_conf = "windows=NO SPACE\nlinux = true;\n";
```
Ditto.

filipegoncalves commented:
The documentation page is showing error 404? It works here. Try the alternative: http://codinghighway.com/rust-config/config/

Anyway, it looks good, I just think we could improve the examples that you changed (and the tests).

And yes, error reporting definitely needs some work (contributions welcome!), I just haven't had the time to get around it yet.

scowcron committed:
Update tests for more clarity.

scowcron commented:
The redirect from github.io worked fine, but the documentation is still showing a 404 for me.

I updated the tests a bit, but there are a lot of ways to make parsing fail. NO SPACE was misleading, I agree. I instead opted to give it a bad boolean value (NOPE). If you'd prefer to see something else there, I'm happy to oblige.

scowcron committed:
Update the grammar in library documentation

scowcron committed:
Upgrade nom to version 1.0

scowcron committed:
Add no terminators tests

scowcron committed:
Fix value, allowing either whitespace, ';', or ','

scowcron committed:
Move reader functions to methods of Config

Rearranged error type a little bit to use the From trait.
Bumped version to 0.2, big interface change.

scowcron committed:
Bump to version 0.2
